server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -91,9 +90,7 @@ public void resync(final IndexShard indexShard, final ActionListener<ResyncTask>
             // Wrap translog snapshot to make it synchronized as it is accessed by different threads through SnapshotSender.
             // Even though those calls are not concurrent, snapshot.next() uses non-synchronized state and is not multi-thread-compatible
             // Also fail the resync early if the shard is shutting down
-            snapshot = indexShard.getHistoryOperations("resync",
-                indexShard.indexSettings.isSoftDeleteEnabled() ? Engine.HistorySource.INDEX : Engine.HistorySource.TRANSLOG,
-                startingSeqNo);
+            snapshot = indexShard.newChangesSnapshot("resync", startingSeqNo, Long.MAX_VALUE, false);
```

Comment:
`readHistoryOperations` used `Math.max(0, startingSeqNo)` as `from`. I wonder if we need to guard against `SequenceNumbers.UNASSIGNED_SEQ_NO` similarly here.

Reply:
I could not find the reason why we used `Math.max(0, startingSeqNo)`. We should always have a valid global checkpoint here, then startingSeqNo should be non-negative. I added an assertion in https://github.com/elastic/elasticsearch/pull/51189/commits/ea8da99f2577a470471f6c2dfc94cb9ccb46b72d.